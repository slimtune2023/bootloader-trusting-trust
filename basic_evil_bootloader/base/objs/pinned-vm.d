objs/pinned-vm.o: pinned-vm.c \
 /Users/jameschen/Documents/cs140e-25win/libpi/include/rpi.h \
 /Users/jameschen/Documents/cs140e-25win/libpi/libc/demand.h \
 /Users/jameschen/Documents/cs140e-25win/libpi/include/gpio.h \
 /Users/jameschen/Documents/cs140e-25win/libpi/include/rpi-constants.h \
 /Users/jameschen/Documents/cs140e-25win/libpi/include/your-prototypes.h \
 /Users/jameschen/Documents/cs140e-25win/libpi/libc/bit-support.h \
 /Users/jameschen/Documents/cs140e-25win/libpi/src/vector-base.h \
 /Users/jameschen/Documents/cs140e-25win/libpi/include/asm-helpers.h \
 pinned-vm.h mmu.h armv6-cp15.h mem-attr.h armv6-coprocessor-asm.h \
 procmap.h memmap-default.h \
 /Users/jameschen/Documents/cs140e-25win/libpi/include/memmap.h

/Users/jameschen/Documents/cs140e-25win/libpi/include/rpi.h:

/Users/jameschen/Documents/cs140e-25win/libpi/libc/demand.h:

/Users/jameschen/Documents/cs140e-25win/libpi/include/gpio.h:

/Users/jameschen/Documents/cs140e-25win/libpi/include/rpi-constants.h:

/Users/jameschen/Documents/cs140e-25win/libpi/include/your-prototypes.h:

/Users/jameschen/Documents/cs140e-25win/libpi/libc/bit-support.h:

/Users/jameschen/Documents/cs140e-25win/libpi/src/vector-base.h:

/Users/jameschen/Documents/cs140e-25win/libpi/include/asm-helpers.h:

pinned-vm.h:

mmu.h:

armv6-cp15.h:

mem-attr.h:

armv6-coprocessor-asm.h:

procmap.h:

memmap-default.h:

/Users/jameschen/Documents/cs140e-25win/libpi/include/memmap.h:
