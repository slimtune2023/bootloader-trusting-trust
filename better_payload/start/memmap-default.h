#ifndef __MEMMAP_DEFAULT_H__
#define __MEMMAP_DEFAULT_H__

// the symbols 
#include "memmap.h"

// we put all the default address space enums here:
// - kernel domain
// - user domain
// - layout.

#define MB(x) ((x)*1024*1024)

// default domains for kernel and user.  not required.
enum {
    dom_kern = 1, // domain id for kernel
    dom_user = 2,  // domain id for user

    // setting for the domain reg: 
    //  - client = checks permissions.
    //  - each domain is 2 bits
    dom_bits = DOM_client << (dom_kern*2) 
        | DOM_client << (dom_user*2) 
};

enum { 
    // default no user access permissions
    no_user = perm_rw_priv,
    // default user access permissions
    user_access = perm_rw_user,

};

// the default asid we use.
enum {
    default_ASID = 1
};


// These are the default segments (segment = one MB)
// that need to be mapped for our binaries so far
// this quarter. 
//
// these will hold for all our tests today.
//
// if we just map these segments we will get faults
// for stray pointer read/writes outside of this region.
enum {
    // code starts at 0x8000, so map the first MB
    //
    // if you look in <libpi/memmap> you can see
    // that all the data is there as well, and we have
    // small binaries, so this will cover data as well.
    //
    // NOTE: obviously it would be better to not have 0 (null) 
    // mapped, but our code starts at 0x8000 and we are using
    // 1mb sections (which require 1MB alignment) so we don't
    // have a choice unless we do some modifications.  
    //
    // you can fix this problem as an extension: very useful!
    SEG_CODE = MB(0),

    // as with previous labs, we initialize 
    // our kernel heap to start at the first 
    // MB. it's 1MB, so fits in a segment. 
    SEG_HEAP = MB(1),

    // if you look in <staff-start.S>, our default
    // stack is at STACK_ADDR, so subtract 1MB to get
    // the stack start.
    SEG_STACK = STACK_ADDR - MB(1),

    // the interrupt stack that we've used all class.
    // (e.g., you can see it in the <full-except-asm.S>)
    // subtract 1MB to get its start
    SEG_INT_STACK = INT_STACK_ADDR - MB(1),

    // the base of the BCM device memory (for GPIO
    // UART, etc).  Three contiguous MB cover it.
    SEG_BCM_0 = 0x20000000,
    SEG_BCM_1 = SEG_BCM_0 + MB(1),
    SEG_BCM_2 = SEG_BCM_0 + MB(2),

    // we guarantee this (2MB) is an 
    // unmapped address
    SEG_ILLEGAL = MB(2),
};

// default kernel attributes
static inline pin_t dev_attr_default(void) {
    return pin_mk_global(dom_kern, no_user, MEM_device);
}
// default kernel attributes
static inline pin_t kern_attr_default(void) {
    return pin_mk_global(dom_kern, no_user, MEM_uncached);
}

// do default identity mapping: should use 16MB

#endif
