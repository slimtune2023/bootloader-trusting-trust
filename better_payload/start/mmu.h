#ifndef __SIMPLE_MMU_H__
#define __SIMPLE_MMU_H__
#include "armv6-cp15.h"
#include "asm-helpers.h"

#define coproc_mk_set(fn_name, coproc, opcode_1, Crn, Crm, opcode_2)       \
    static inline void c ## coproc ## _ ## fn_name ## _set(uint32_t v) {                    \
        asm volatile ("mcr " MK_STR(coproc) ", "                        \
                             MK_STR(opcode_1) ", "                      \
                             "%0, "                                     \
                            MK_STR(Crn) ", "                            \
                            MK_STR(Crm) ", "                            \
                            MK_STR(opcode_2) :: "r" (v));               \
        prefetch_flush();                                               \
    }

#define coproc_mk_get(fn_name, coproc, opcode_1, Crn, Crm, opcode_2)       \
    static inline uint32_t c ## coproc ## _ ## fn_name ## _get(void) {                      \
        uint32_t ret=0;                                                   \
        asm volatile ("mrc " MK_STR(coproc) ", "                        \
                             MK_STR(opcode_1) ", "                      \
                             "%0, "                                     \
                            MK_STR(Crn) ", "                            \
                            MK_STR(Crm) ", "                            \
                            MK_STR(opcode_2) : "=r" (ret));             \
        return ret;                                                     \
    }

#define coproc_mk(fn, coproc, opcode_1, Crn, Crm, opcode_2)     \
coproc_mk_set(fn, coproc, opcode_1, Crn, Crm, opcode_2)        \
coproc_mk_get(fn, coproc, opcode_1, Crn, Crm, opcode_2) 

// low level MMU hardware routines.  agnostic
// as to whether we use a page table or pinning.
//
// workflow:
//  1. init mmu before doing anything.
//  2. then setup:
//   - domain permissions
//   - the asid
//   - page table (invalid or valid)
//   - any pinned entries.
// 3. then can turn the mmu on.
// 4. then can read/write memory.
// 5. then can turn the mmu off.
//
// NOTE: there are generally single hardware instructions
// for each of the above, HOWEVER, just issuing the 
// instruction is rarely enough. there is typically
// a recipe to follow to ensure that hardware state
// is consistent.  See the upcoming vm coherence lab.

// One time hardware initialization.  
// Do before anything else!
void staff_mmu_init(void);
void mmu_init(void);

coproc_mk(dom, p15, 0, c3, c0, 0)

// Set the domain access control to <d>.
// do before turning MMU on!
void staff_domain_access_ctrl_set(uint32_t d);
void domain_access_ctrl_set(uint32_t d);

// get
uint32_t staff_domain_access_ctrl_get(void);
uint32_t domain_access_ctrl_get(void);


// internal routine to set the hardware state: 
//   - <asid> is address space identifier.
//   - <pt> is the 2^14 aligned page table.
//
// do before turning MMU on!
void staff_set_procid_ttbr0(unsigned pid, unsigned asid, void *pt);
void set_procid_ttbr0(unsigned pid, unsigned asid, void *pt);

// called to sync up the hw state after modifying tlb entry
void staff_sync_tlb(void);

// setup pid, asid and pt in hardware.
// must call:
//  1. before turning MMU on at all
//  2. when switching address spaces (or asid won't
//     be correct).
void staff_mmu_set_ctx(uint32_t pid, uint32_t asid, void *pt);
void mmu_set_ctx(uint32_t pid, uint32_t asid, void *pt);


// called to sync after a set of pte modifications: flushes everything.
void mmu_sync_pte_mods(void);
void staff_mmu_sync_pte_mods(void);



#if 0
static inline void 
staff_mmu_set_ctx(uint32_t pid, uint32_t asid, void *pt) {
    assert(asid!=0);
    assert(asid<64);
    staff_set_procid_ttbr0(pid, asid, pt);
}

static inline void 
mmu_set_ctx(uint32_t pid, uint32_t asid, void *pt) {
    assert(asid!=0);
    assert(asid<64);
    set_procid_ttbr0(pid, asid, pt);
}
#endif

// turn the MMU on: 
// you must have previously done:
//  - <mmu_init> or hardware could have garbage.
//  - <mmu_set_ctx>: or there is no asid and page 
//    table for the hw to use.
//  - <domain_access_control_set> or no domain
//    permissions will be setup and you will immediately
//    fault.
void staff_mmu_enable(void);
void mmu_enable(void);

// turn mmu off, write-back data cache if needed, etc.
void staff_mmu_disable(void);
void mmu_disable(void);

// helper routine: is mmu on?
int staff_mmu_is_enabled(void);
int mmu_is_enabled(void);

void reset(void);
void mmu_disable_set_asm(cp15_ctrl_reg1_t c);
void mmu_enable_set_asm(cp15_ctrl_reg1_t c);
void mmu_sync_pte_mods(void);
void cp15_domain_ctrl_wr(uint32_t dom_reg);


#endif
