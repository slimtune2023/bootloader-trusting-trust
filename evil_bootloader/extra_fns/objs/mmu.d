objs/mmu.o: mmu.c \
 /Users/jameschen/Documents/cs140e-25win/libpi/include/rpi.h \
 /Users/jameschen/Documents/cs140e-25win/libpi/libc/demand.h \
 /Users/jameschen/Documents/cs140e-25win/libpi/include/gpio.h \
 /Users/jameschen/Documents/cs140e-25win/libpi/include/rpi-constants.h \
 /Users/jameschen/Documents/cs140e-25win/libpi/include/your-prototypes.h \
 /Users/jameschen/Documents/cs140e-25win/libpi/include/rpi-constants.h \
 /Users/jameschen/Documents/cs140e-25win/libpi/include/rpi-interrupts.h \
 /Users/jameschen/Documents/cs140e-25win/libpi/libc/helper-macros.h \
 mmu-internal.h armv6-cp15.h mmu.h \
 /Users/jameschen/Documents/cs140e-25win/libpi/include/asm-helpers.h

/Users/jameschen/Documents/cs140e-25win/libpi/include/rpi.h:

/Users/jameschen/Documents/cs140e-25win/libpi/libc/demand.h:

/Users/jameschen/Documents/cs140e-25win/libpi/include/gpio.h:

/Users/jameschen/Documents/cs140e-25win/libpi/include/rpi-constants.h:

/Users/jameschen/Documents/cs140e-25win/libpi/include/your-prototypes.h:

/Users/jameschen/Documents/cs140e-25win/libpi/include/rpi-constants.h:

/Users/jameschen/Documents/cs140e-25win/libpi/include/rpi-interrupts.h:

/Users/jameschen/Documents/cs140e-25win/libpi/libc/helper-macros.h:

mmu-internal.h:

armv6-cp15.h:

mmu.h:

/Users/jameschen/Documents/cs140e-25win/libpi/include/asm-helpers.h:
