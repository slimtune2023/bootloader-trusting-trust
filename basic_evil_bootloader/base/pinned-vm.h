#ifndef __PINNED_VM_H__
#define __PINNED_VM_H__
// simple interface for pinned virtual memory: most of it is 
// enum and data structures.  you'll build three routines:

#include "mmu.h"
#include "mem-attr.h"
#include "asm-helpers.h"
#include "armv6-coprocessor-asm.h"

// you can flip these back and forth if you want debug output.
#if 0
    // change <output> to <debug> if you want file/line/func
#   define pin_debug(args...) output("PIN_VM:" args)
#else
#   define pin_debug(args...) do { } while(0)
#endif

#define coproc_mk_set(fn_name, coproc, opcode_1, Crn, Crm, opcode_2)       \
    static inline void c ## coproc ## _ ## fn_name ## _set(uint32_t v) {                    \
        asm volatile ("mcr " MK_STR(coproc) ", "                        \
                             MK_STR(opcode_1) ", "                      \
                             "%0, "                                     \
                            MK_STR(Crn) ", "                            \
                            MK_STR(Crm) ", "                            \
                            MK_STR(opcode_2) :: "r" (v));               \
        prefetch_flush();                                               \
    }

#define coproc_mk_get(fn_name, coproc, opcode_1, Crn, Crm, opcode_2)       \
    static inline uint32_t c ## coproc ## _ ## fn_name ## _get(void) {                      \
        uint32_t ret=0;                                                   \
        asm volatile ("mrc " MK_STR(coproc) ", "                        \
                             MK_STR(opcode_1) ", "                      \
                             "%0, "                                     \
                            MK_STR(Crn) ", "                            \
                            MK_STR(Crm) ", "                            \
                            MK_STR(opcode_2) : "=r" (ret));             \
        return ret;                                                     \
    }


// make both get and set methods.
#define coproc_mk(fn, coproc, opcode_1, Crn, Crm, opcode_2)     \
    coproc_mk_set(fn, coproc, opcode_1, Crn, Crm, opcode_2)        \
    coproc_mk_get(fn, coproc, opcode_1, Crn, Crm, opcode_2) 

coproc_mk(tlb_idx, p15, 5, c15, c4, 2)
coproc_mk(tlb_va, p15, 5, c15, c5, 2)
coproc_mk(tlb_pa, p15, 5, c15, c6, 2)
coproc_mk(tlb_attr, p15, 5, c15, c7, 2)
coproc_mk_set(va_trans, p15, 0, c7, c8, 0) // op code 2?
coproc_mk_get(pa_trans, p15, 0, c7, c4, 0)

// attributes: these get inserted into the TLB.
typedef struct {
    // for today we only handle 1MB sections.
    uint32_t G,         // is this a global entry?
             asid,      // don't think this should actually be in this.
             dom,       // domain id
             pagesize;  // can be 1MB or 16MB

    // permissions for needed to access page.
    //
    // see mem_perm_t above: is the bit merge of 
    // APX and AP so can or into AP position.
    mem_perm_t  AP_perm;

    // caching policy for this memory.
    // 
    // see mem_cache_t enum above.
    // see table on 6-16 b4-8/9
    // for today everything is uncacheable.
    mem_attr_t mem_attr;
} pin_t;

// pinned encodings for different page sizes.
enum {
    PAGE_4K     = 0b01,
    PAGE_64K    = 0b10,
    PAGE_1MB    = 0b11,
    PAGE_16MB   = 0b00
};

static inline pin_t pin_16mb(pin_t p) {
    p.pagesize = PAGE_16MB;
    return p;
}

// enable MMU -- must have set the context
// first.
void pin_mmu_enable(void);

// disable MMU
void pin_mmu_disable(void);

// mmu can be off or on: setup address space context.
// must do before enabling MMU or when switching
// between address spaces.
void pin_set_context(uint32_t asid);

// our main routine: 
//  insert map <va> ==> <pa> with attributes <attr> 
//  at position <idx> in the TLB.
//
// errors:
//  - idx >= 8.
//  - va already mapped.
void pin_mmu_sec(unsigned idx,
                uint32_t va,
                uint32_t pa,
                pin_t attr);

// do a manual translation in tlb and see if exists (1)
// returns the result in <result>
//
// low 1..6 bits of result should have the reason.
int tlb_contains_va(uint32_t *result, uint32_t va);

// wrapper that check that <va> is pinned in the tlb
int pin_exists(uint32_t va, int verbose_p);

// print <msg> then all valid pinned entries.
// note: if you print everything you see a lot of
// garbage in the non-initialized entires --- i'm 
// not sure if we are guaranteed that these have their
// valid bit set to 0?   
void lockdown_print_entries(const char* msg);

// set pin index <idx> to all 0s.
void pin_clear(unsigned idx);

void pin_mmu_init(uint32_t domain_reg);

// void staff_pin_mmu_init(uint32_t domain_reg);

static inline pin_t pin_mk_global(uint32_t dom, mem_perm_t perm, mem_attr_t attr) {
    pin_t p;
    p.G = 1;
    p.asid = 0;
    p.dom = dom;
    p.pagesize = 0b11;
    p.AP_perm = perm;
    p.mem_attr = attr;
    return p;
}

static inline pin_t pin_mk_user(uint32_t dom, uint32_t asid, mem_perm_t perm, mem_attr_t attr) {
    pin_t p;
    p.G = 0;
    p.asid = asid;
    p.dom = dom;
    p.pagesize = 0b11;
    p.AP_perm = perm;
    p.mem_attr = attr;
    return p;
}

static inline pin_t pin_mk_device(uint32_t dom) {
    pin_t p;
    p.G = 1;
    p.asid = 0;  // unset
    p.dom = dom;
    p.pagesize = 0b11;
    p.AP_perm = perm_rw_priv;
    p.mem_attr = MEM_device;
    return p;
}

#endif