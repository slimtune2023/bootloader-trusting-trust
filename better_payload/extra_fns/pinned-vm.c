#include "rpi.h"
#include "libc/bit-support.h"

// has useful enums and helpers.
#include "vector-base.h"
#include "pinned-vm.h"
#include "mmu.h"
#include "procmap.h"

// generate the _get and _set methods.
// (see asm-helpers.h for the cp_asm macro 
// definition)
// arm1176.pdf: 3-149

// static void *null_pt = 0;

void pin_mmu_enable(void) {
    mmu_enable();
}

void pin_mmu_disable(void) {
    mmu_disable();
}

void pin_set_context(uint32_t asid) {
    void *null_pt = kmalloc_aligned(4096*4, 1<<14);
    mmu_set_ctx(128, asid, null_pt);
}

// our main routine: 
//  insert map <va> ==> <pa> with attributes <attr> 
//  at position <idx> in the TLB.
//
// errors:
//  - idx >= 8.
//  - va already mapped.
void pin_mmu_sec(unsigned idx, uint32_t va, uint32_t pa, pin_t attr) {
    assert(idx < 8);
    uint32_t *temp = 0;
    // assert(!tlb_contains_va(temp, va));

    cp15_tlb_idx_set(idx);


    // printk("asid: %x\n", attr.asid);
    unsigned va_set = attr.asid;
    va_set |= attr.G << 9;
    va_set |= va;
    cp15_tlb_va_set(va_set);

    unsigned attr_set = attr.mem_attr << 1;
    attr_set |= attr.dom << 7;
    cp15_tlb_attr_set(attr_set);

    unsigned pa_set = 1;
    pa_set |= attr.AP_perm << 1;
    pa_set |= attr.pagesize << 6; // size; change if not 1mb
    pa_set |= 0b00 << 8;
    pa_set |= pa;
    cp15_tlb_pa_set(pa_set);

    asm volatile ("mcr p15, 0, r0, c7, c10, 4");  // DSB
    asm volatile ("mcr p15, 0, r0, c7, c5, 4");   // prefetch_flush

    // assert(tlb_contains_va(temp, va));
    // branch thingy?
}

// do a manual translation in tlb and see if exists (1)
// returns the result in <result>
//
// low 1..6 bits of result should have the reason.
int tlb_contains_va(uint32_t *result, uint32_t va) {
    cp15_va_trans_set(va);
    *result = cp15_pa_trans_get();
    uint32_t ret = 1-(*result & 1);
    *result &= ~0x3ff;
    *result |= (va & 0x3ff);
    return ret;
}

int pin_exists(uint32_t va, int verbose_p) {
    uint32_t res;
    return tlb_contains_va(&res, va);
}

uint32_t* get_regs_idx(unsigned idx) {
    cp15_tlb_idx_set(idx);

    unsigned va = cp15_tlb_va_get();
    unsigned pa = cp15_tlb_pa_get();
    unsigned attr = cp15_tlb_attr_get();

    asm volatile ("mcr p15, 0, r0, c7, c10, 4");  // DSB
    asm volatile ("mcr p15, 0, r0, c7, c5, 4");   // prefetch_flush

    uint32_t* ret = 0;
    *ret = va;
    *(ret+1) = pa;
    *(ret+2) = attr;
    return ret;
}

// void lockdown_print_entries(const char* msg) {
//     printk("%s\n", msg);
//     unsigned* ret;
//     uint32_t* res = 0;
//     for (int i = 0; i < 8; i++) {
//         ret = get_regs_idx(i);
//         printk("TLB entry %d:\n", i);
//         printk("va: %x\t", ret[0]);
//         printk("pa: %x\t", ret[1]);
//         printk("attr: %x\t", ret[2]);
//         tlb_contains_va(res, ret[0]);
//         printk("contains: %x\n", *res);
//     }
// }

void lockdown_print_entry(unsigned idx) {
    trace("   idx=%d\n", idx);
    // lockdown_index_set(idx);
    // uint32_t va_ent = lockdown_va_get();
    // uint32_t pa_ent = lockdown_pa_get();

    uint32_t *ret = get_regs_idx(idx);
    uint32_t va_ent = ret[0];
    uint32_t pa_ent = ret[1];
    uint32_t attr = ret[2];

    uint32_t v = pa_ent & 0x1; 

    if(!v) {
        trace("     [invalid entry %d]\n", idx);
        return;
    }

    // 3-149
    uint32_t va = (va_ent >> 12);
    uint32_t G = (va_ent >> 9) & 1;
    uint32_t asid = va_ent & 0xff;
    trace("     va_ent=%x: va=%x|G=%d|ASID=%d\n",
        va_ent, va, G, asid);

    // 3-150
    uint32_t pa = (pa_ent >> 12);
    uint32_t nsa = (pa_ent >> 9) & 1;
    uint32_t nstid = (pa_ent >> 8) & 1;
    uint32_t size = (pa_ent >> 6) & 0x3; 
    uint32_t apx = (pa_ent >> 1) & 0x3;
    trace("     pa_ent=%x: pa=%x|nsa=%d|nstid=%d|size=%b|apx=%b|v=%d\n",
                pa_ent, pa, nsa,nstid,size, apx,v);

    // 3-151
    uint32_t dom = (attr >> 7) & 0xf;
    uint32_t xn = (attr >> 6) & 1;
    uint32_t tex = (attr >> 3) & 0x7; 
    uint32_t C = (attr >> 2) & 0x1;
    uint32_t B = (attr >> 1) & 0x1;  
    trace("     attr=%x: dom=%d|xn=%d|tex=%b|C=%d|B=%d\n",
            attr, dom,xn,tex,C,B);
}

void lockdown_print_entries(const char *msg) {
    trace("-----  <%s> ----- \n", msg);
    trace("  pinned TLB lockdown entries:\n");
    for(int i = 0; i < 8; i++)
        lockdown_print_entry(i);
    trace("----- ---------------------------------- \n");
}

void pin_clear(unsigned idx) {
    return;
}

void pin_mmu_init(uint32_t domain_reg) {
    mmu_init();
    domain_access_ctrl_set(domain_reg);
}

