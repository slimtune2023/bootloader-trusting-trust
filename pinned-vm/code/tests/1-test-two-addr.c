// test that we can:
//   1. create private virt->phys mappings
//   2. switch between virtual address spaces
// by:
//   1. create two address spaces (ASID1, ASID2)
//   2. map the same virtual address <user_addr> in both
//      to different physical segments (<phys_addr1> and
//      <phys_addr2>.
//   3. switch between them reading and writing
#include "rpi.h"
#include "pinned-vm.h"
#include "mmu.h"
#include "memmap-default.h"
#include "full-except.h"

void notmain(void) { 
    // map the heap: for lab cksums must be at 0x100000.
    kmalloc_init_set_start((void*)MB(1), MB(1));
    full_except_install(0);
    assert(!mmu_is_enabled());

    // default domain bits.
    pin_mmu_init(dom_bits);

    unsigned idx = 0;
    pin_t kern = pin_mk_global(dom_kern, no_user, MEM_uncached);
    pin_mmu_sec(idx++, SEG_CODE, SEG_CODE, kern);
    pin_mmu_sec(idx++, SEG_HEAP, SEG_HEAP, kern);
    pin_mmu_sec(idx++, SEG_STACK, SEG_STACK, kern);
    pin_mmu_sec(idx++, SEG_INT_STACK, SEG_INT_STACK, kern);

    // use 16mb section for device.
    pin_t dev  = pin_16mb(pin_mk_global(dom_kern, no_user, MEM_device));
    pin_mmu_sec(idx++, SEG_BCM_0, SEG_BCM_0, dev);

    enum { ASID1 = 1, ASID2 = 2 };
    // do a non-ident map
    enum {
        user_addr = MB(16),
        phys_addr1 = user_addr+MB(1),
        phys_addr2 = user_addr+MB(2)
    };
    pin_t user1 = pin_mk_user(dom_kern, ASID1, no_user, MEM_uncached);
    pin_t user2 = pin_mk_user(dom_kern, ASID2, no_user, MEM_uncached);
    pin_mmu_sec(idx++, user_addr, phys_addr1, user1);
    pin_mmu_sec(idx++, user_addr, phys_addr2, user2);

    // initialize with MMU off.
    PUT32(phys_addr1, 0x11111111);
    PUT32(phys_addr2, 0x22222222);

    assert(idx<8);

    trace("about to enable\n");
    lockdown_print_entries("about to turn on first time");

    // *****************************************************
    // turn on address space 1 and check we can write.

    pin_set_context(ASID1);
    pin_mmu_enable();
        trace("MMU is on and working!\n");
    
        uint32_t x = GET32(user_addr);
        trace("ASID %d = got: %x\n", ASID1, x);
        assert(x == 0x11111111);
        PUT32(user_addr, ASID1);

    pin_mmu_disable();

    // check that the write happened.
    trace("phys addr1=%x\n", GET32(phys_addr1));
    assert(GET32(phys_addr1) == ASID1);


    // *****************************************************
    // switch to 2nd address space and check we can write.
    pin_set_context(ASID2);
    mmu_enable();
        x = GET32(user_addr);
        trace("asid %d = got: %x\n", ASID2, x);
        assert(x == 0x22222222);
        PUT32(user_addr, ASID2);
    pin_mmu_disable();

    trace("phys addr2=%x\n", GET32(phys_addr2));
    assert(GET32(phys_addr2) == ASID2);

    // *****************************************************
    // now check that works by switching even with MMU on.

    trace("about to check that can switch ASID w/ MMU on\n");
    PUT32(phys_addr1, 0x11111111); // reset 
    PUT32(phys_addr2, 0x22222222);

    pin_set_context(ASID1);
    pin_mmu_enable();
    assert(GET32(user_addr) == 0x11111111);
    PUT32(user_addr, ASID1);

    pin_set_context(ASID2);
    assert(GET32(user_addr) == 0x22222222);
    PUT32(user_addr, ASID2);

    pin_mmu_disable();

    // make sure that the writes happened.
    assert(GET32(phys_addr1) == ASID1);
    assert(GET32(phys_addr2) == ASID2);
    trace("SUCCESS!\n");
}
