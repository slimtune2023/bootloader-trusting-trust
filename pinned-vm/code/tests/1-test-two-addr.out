TRACE:notmain:about to enable
TRACE:lockdown_print_entries:-----  <about to turn on first time> ----- 
TRACE:lockdown_print_entries:  pinned TLB lockdown entries:
TRACE:lockdown_print_entry:   idx=0
TRACE:lockdown_print_entry:     va_ent=0x200: va=0x0|G=1|ASID=0
TRACE:lockdown_print_entry:     pa_ent=0xc3: pa=0x0|nsa=0|nstid=0|size=11|apx=1|v=1
TRACE:lockdown_print_entry:     attr=0x88: dom=1|xn=0|tex=1|C=0|B=0
TRACE:lockdown_print_entry:   idx=1
TRACE:lockdown_print_entry:     va_ent=0x100200: va=0x100|G=1|ASID=0
TRACE:lockdown_print_entry:     pa_ent=0x1000c3: pa=0x100|nsa=0|nstid=0|size=11|apx=1|v=1
TRACE:lockdown_print_entry:     attr=0x88: dom=1|xn=0|tex=1|C=0|B=0
TRACE:lockdown_print_entry:   idx=2
TRACE:lockdown_print_entry:     va_ent=0x7f00200: va=0x7f00|G=1|ASID=0
TRACE:lockdown_print_entry:     pa_ent=0x7f000c3: pa=0x7f00|nsa=0|nstid=0|size=11|apx=1|v=1
TRACE:lockdown_print_entry:     attr=0x88: dom=1|xn=0|tex=1|C=0|B=0
TRACE:lockdown_print_entry:   idx=3
TRACE:lockdown_print_entry:     va_ent=0x8f00200: va=0x8f00|G=1|ASID=0
TRACE:lockdown_print_entry:     pa_ent=0x8f000c3: pa=0x8f00|nsa=0|nstid=0|size=11|apx=1|v=1
TRACE:lockdown_print_entry:     attr=0x88: dom=1|xn=0|tex=1|C=0|B=0
TRACE:lockdown_print_entry:   idx=4
TRACE:lockdown_print_entry:     va_ent=0x20000200: va=0x20000|G=1|ASID=0
TRACE:lockdown_print_entry:     pa_ent=0x20000003: pa=0x20000|nsa=0|nstid=0|size=0|apx=1|v=1
TRACE:lockdown_print_entry:     attr=0x80: dom=1|xn=0|tex=0|C=0|B=0
TRACE:lockdown_print_entry:   idx=5
TRACE:lockdown_print_entry:     va_ent=0x1000001: va=0x1000|G=0|ASID=1
TRACE:lockdown_print_entry:     pa_ent=0x11000c3: pa=0x1100|nsa=0|nstid=0|size=11|apx=1|v=1
TRACE:lockdown_print_entry:     attr=0x88: dom=1|xn=0|tex=1|C=0|B=0
TRACE:lockdown_print_entry:   idx=6
TRACE:lockdown_print_entry:     va_ent=0x1000002: va=0x1000|G=0|ASID=2
TRACE:lockdown_print_entry:     pa_ent=0x12000c3: pa=0x1200|nsa=0|nstid=0|size=11|apx=1|v=1
TRACE:lockdown_print_entry:     attr=0x88: dom=1|xn=0|tex=1|C=0|B=0
TRACE:lockdown_print_entry:   idx=7
TRACE:lockdown_print_entry:     
TRACE:lockdown_print_entries:----- ---------------------------------- 
TRACE:notmain:MMU is on and working!
TRACE:notmain:ASID 1 = got: 0x11111111
TRACE:notmain:phys addr1=0x1
TRACE:notmain:asid 2 = got: 0x22222222
TRACE:notmain:phys addr2=0x2
TRACE:notmain:about to check that can switch ASID w/ MMU on
TRACE:notmain:SUCCESS!
